odoo/addons/base/models/ir_cron.py

```diff
@@ -133,12 +132,28 @@ def method_direct_trigger(self):
         job = self._acquire_one_job(cron_cr, self.id, include_not_ready=True)
         if not job:
             raise UserError(self.env._("Job '%s' already executing", self.name))
-        self._process_job(cron_cr, job)
+        if (limit := config['limit_time_real']) > 0:
+            # halve the available hard limit
+            end_time = time.monotonic() + limit // 2
+        else:
+            end_time = None
+        self._process_job(cron_cr, job, end_time=end_time)
         return True
 
     @staticmethod
-    def _process_jobs(db_name: str) -> None:
+    def _process_jobs(db_name: str, soft_limit: float | None = None) -> None:
         """ Execute every job ready to be run on this database. """
+        if soft_limit is None:
+            soft_limit = config['limit_time_soft_cron']
+            if soft_limit < 0:
+                # default to half of the hard-limit
+                real_limit_cron = config['limit_time_real_cron']
+                if real_limit_cron < 0:
+                    real_limit_cron = config['limit_time_real']
+                soft_limit = (real_limit_cron + 1) // 2
```

Comment:
I wonder if this bit of code couldn't be placed inside `config.py`, to expose inside `runtime_config` the result of the fallback, should `config['limit_time_soft_cron']` be `-1`.
So we wouldn't need to have this computation everything we wanna use the config value (granted, at the moment we only use it here).

Reply:
There are some other places that do similar logic. They should all be refactored at the same time.